### ISSUE TITLE
Signed Hash Verification Signature has extra bytes which appear to be meaningless

### ISSUE THREAD

**Issue by alrossi** [xrootd/xrootd#1046]
I am opening this issue here because the discussion on the dev group concerning this is stalled.

From the point of view of the current dCache implementation, there seems to be a length mismatch between the actual signature sent and the hashing of the message, even though the hashed part corresponds.

If I understand correctly, when the message header + body is hashed using SHA256, the result should be a 32 byte sequence.

The signature that the dCache server is receiving from the 4.10.0 client, however, is 48 bytes long.   Moreover, the extra 16 bytes consistently decode to:

```
0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10 0x10
```

which seems to serve no purpose.

To illustrate further, here is a TRACE level logging excerpt from the dCache server:

![thumbnail_example](https://user-images.githubusercontent.com/3985564/63521795-bae55c00-c4bc-11e9-9a8d-138fa239c02b.png)

The first byte dump shows the encrypted SigVer signature hash.   The prepended IV vector (red rectangle) is extracted, and decryption proceeds as normal, producing the decrypted version which consists of the green rectangle (hash proper), plus the blue 16-byte sequence above.

Notice then that the actual message (in this case, a kXR_stat request), when hashed according to the protocol specification –– that is, using the received random 64-bit sequence number + header + payload –– produces exactly the byte sequence in the green rectangle.

However, the comparison fails because the hash is 32 bytes long, and missing the 16 byte nonce.

If I change the code to ignore any bytes in the SigVer signature after the 32nd, then the comparison works.

It is not clear to me why those extra 16 bytes are there, and what I should do about them.

Nor is it clear whether the symmetrical encryption part, which dCache would use to sign requests from its third-party client to send to the xrootd server, should be similarly expanded by 16 bytes.  And why those bytes, before encryption, should be uniformly 0x10 (i.e., decimal 16).

Thank you for looking at this and helping me clarify.

Al

**Comment by alrossi** [comment]
```Nor is it clear whether the symmetrical encryption part, which dCache would use to sign requests from its third-party client to send to the xrootd server, should be similarly expanded by 16 bytes. And why those bytes, before encryption, should be uniformly 0x10 (i.e., decimal 16).```

I have gone ahead and changed the encryption as well in order to make dCache compatible with the 4.10 server.  Seems the same problem does indeed exist symmetrically.

I have a patch to restore compatibility, 

https://github.com/alrossi/xrootd4j/tree/fix/master/dh-signature

but it would be good to know why those extra bytes are there.

**Comment by abh3** [comment]
Have we resolved this?

**Comment by alrossi** [comment]
My solution was as I stated above:  I just ignore the extra bytes.   I don't know whether that can be considered a real solution or not, since I still do not understand why those extra bytes are there.

**Comment by wyang007** [comment]
@alrossi It looks like there are two things here. In the C++ Xrootd server (https://github.com/xrootd/xrootd/blob/master/src/XrdCrypto/XrdCryptosslCipher.cc#L1091), EVP_CipherUpdate() take a buffer with length of 48, and produce a decrypted hash of 32 bytes (so that 16 bytes in the second rectangular is not produced in C++ server!). This seems to be different from what you observer. So could this be the difference of C++ and Java implementation? 

Also, I can set the buffer length to anything between 32 to 48 and it produce the same result. Client will get an error if I set it to 31 bytes. So this does indicate that the 16 byte in your first blue rectangular is not needed.

**Comment by abh3** [comment]
Huh? I a totally lost! Please provide a lot more context!

**Comment by wyang007** [comment]
btw, the initial string of IV(16byte)+Encrypted Hash(32byte)+SomethingElse(16byte) came from XrdSecProtect::Verify()

**Comment by xrootd-dev** [comment]
Does this mean the superflous bytes are not needed? If so, can we get id o 
them to avoid the ambiguity? Am I even on the right discussion here?

On Tue, 14 Apr 2020, Wei Yang wrote:

c> @alrossi It looks like there are two things here. In the C++ Xrootd 
server 
(https://github.com/xrootd/xrootd/blob/master/src/XrdCrypto/XrdCryptosslCipher.cc#L1091), EVP_CipherUpdate() take a buffer with length of 48, and produce a decrypted hash of 32 bytes (so that 16 bytes in the second rectangular is not produced in C++ server!). This seems to be different from what you observer. So could this be the difference of C++ and Java implementation?

**Comment by abh3** [comment]
So, what do I need to correct this?

**Comment by alrossi** [comment]
I haven't looked at the C++ implementation in a while, but I am not sure why it is using a 48 byte buffer, but it seems to me that only the first 32 bytes of the buffer are "real" and the rest is padding.    In any event, if what Wei points out is correct, then you must be doing internally something similar to what dCache needs to do when it receives the hash.

The question I suppose that needs to be examined is why the buffer is 48 bytes to begin with.

**Comment by wyang007** [comment]
(In e-mail, last paragraph from @alrossi is missing! Not sure what was going on but I just want to bring this up because this could be a serious problem with github notification)

The behavior of EVP_CipherUpdate() basically filter out the extra 16 bytes. So that is why we never see this problem in C++ implementation. I agree we need to find out where this 16 extra bytes coming from.

**Comment by alrossi** [comment]
Wei, I think I added the comment after the email.

Is EVP_CipherUpdate() an xrootd method or is it an openssl library call?

If the latter, then it would seem that in general one is warranted in truncating anything longer than 32 bytes to 32 bytes, in which case, dCache is simply doing what the openssl library does.

This would suggest to me that the library itself does not expect to receive exactly 32 bytes, and so neither should dCache.

If this is correct, then I guess we can consider this a non-issue.

**Comment by wyang007** [comment]
@alrossi EVP_CipherUpdate() is an OpenSSL library function.

**Comment by alrossi** [comment]
OK.   But just to clarify:  

(a) xrdcp sends a 48-byte message.
(b) the C++ server turns this into 32 bytes (truncates) using OpenSSL.
(c) dCache turns this into 32 bytes (truncates) similarly.

At this point, if the C++ server has to do the same thing as the dCache server, then you may feel  free to close this issue.

I would think, though, that the client should be changed so that it emits 32 bytes, not 48, since the last 16 are meaningless.  Less wasteful over-the-wire packets.

Thanks, Al

**Comment by wyang007** [comment]
indeed, we need to find out the origin of that 16 wasteful bytes. It looks like you already confirmed that it was coming from the client (instead of being generated by some buffer at the receiving end).

**Comment by alrossi** [comment]
My guess is that it is simply an over-allocated but properly initialized buffer (i.e., starts out with 48 0\s).  First 32 bytes are written to, but the entire buffer is then written to the "bucket".

**Comment by wyang007** [comment]
@alrossi I took a closer look at https://github.com/xrootd/xrootd/blob/master/src/XrdCrypto/XrdCryptosslCipher.cc#L1091 (from line 1091 to 1103. There is an EVP_CipherUpdate() (which set ltmp=32) and an EVP_CipherFinal_ex() call (which set ltmp=16). And the final returned value is 48 (the length of the buffer). The last 16 bytes are those that got transmitted over the wire.

I didn't find document to explain in detail about these two functions. But code example shows exactly the same way we use them in XrdCryptosslCipher.cc. So I guess this is just how the C++ Openssl works (I also tried to return 32 instead of 48 but that cause failure in an earlier step)
